sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+      }
+      return (ParameterizedType) genericSuperclass;
+    }
+
+    /**
+     * Returns RecordReader object of the first split to read first record for validating key/value
+     * classes.
+     */
+    private RecordReader fetchFirstRecord() throws IOException, InterruptedException {
+      RecordReader<?, ?> reader =
+          inputFormatObj.createRecordReader(inputSplits.get(0).getSplit(), taskAttemptContext);
+      if (reader == null) {
+        throw new IOException(
+            String.format(HadoopInputFormatIOConstants.NULL_CREATE_RECORDREADER_ERROR_MSG,
+                inputFormatObj.getClass()));
+      }
+      reader.initialize(inputSplits.get(0).getSplit(), taskAttemptContext);
+      // First record is read to get the InputFormat's key and value classes.
+      reader.nextKeyValue();
+      return reader;
+    }
+
+    @VisibleForTesting
+    InputFormat<?, ?> getInputFormat(){
+      return inputFormatObj;
+    }
+
+    @VisibleForTesting
+    void setInputFormatObj(InputFormat<?, ?> inputFormatObj) {
+      this.inputFormatObj = inputFormatObj;
+    }
+
+    @Override
+    public Coder<KV<K, V>> getDefaultOutputCoder() {
+      return KvCoder.of(keyCoder, valueCoder);
+    }
+
+    private Class<?> getExpectedKeyClass() {
+      return expectedKeyClass;
+    }
+
+    private void setExpectedKeyClass(Class<?> expectedKeyClass) {
+      this.expectedKeyClass = expectedKeyClass;
+    }
+
+    private Class<?> getExpectedValueClass() {
+      return expectedValueClass;
+    }
+
+    private void setExpectedValueClass(Class<?> expectedValueClass) {
+      this.expectedValueClass = expectedValueClass;
+    }
+
+    @Override
+    public BoundedReader<KV<K, V>> createReader(PipelineOptions options) throws IOException {
+      this.validate();
+      if (inputSplit == null) {
+          throw new IOException(HadoopInputFormatIOConstants.CREATEREADER_UNSPLIT_SOURCE_ERROR_MSG);
+      } else {
+        createInputFormatInstance();
+        return new HadoopInputFormatReader<>(
+            this,
+            keyTranslationFunction,
+            valueTranslationFunction,
+            inputSplit,
+            inputFormatObj,
+            taskAttemptContext);
+      }
+    }
+
+    /**
+     * BoundedReader for Hadoop InputFormat source.
+     *
+     * @param <K> Type of keys RecordReader emits.
+     * @param <V> Type of values RecordReader emits.
+     */
+    class HadoopInputFormatReader<T1, T2> extends BoundedSource.BoundedReader<KV<K, V>> {
+
+      private final HadoopInputFormatBoundedSource<K, V> source;
+      @Nullable private final SimpleFunction<T1, K> keyTranslationFunction;
+      @Nullable private final SimpleFunction<T2, V> valueTranslationFunction;
+      private final SerializableSplit split;
+      private RecordReader<T1, T2> currentReader;
+      private volatile boolean doneReading = false;
+      private long recordsReturned = 0L;
+      // Variable added to track the progress of the RecordReader.
```

Comment:
is this comment really adding value? If you think we need a comment, I wouldn't phrase it as "Variable added to..." - just comment on what it does.